### PR TITLE
[aws-lambda] Some docs, split up types and tests

### PR DESCRIPTION
**Pull request by simonbuchan**
Documents handler type and conventions for adding triggers more
explicitly. The same index.d.ts docs should explain the thinking for the
new layout convention.

No changes to the type tests in this first commit, to show that the tests
still passs, and because I think they deserve more attention (and I need
lunch right now).

Most event structures have little to no changes other than re-ordering or removing
inaccurately copy/pasted comments in a case or two.

Loses history, but github doesn't support rename history anyway, and
it's even more ugly to get git to track file splits even for local
usage. Apologies for the huge diff.

Note to DT maintainer: `npm test` is failing locally for me due to `Error: Unused file C:\code\personal\DefinitelyTyped/types/react-pdf/OTHER_FILES.txt` - not sure if there's a good way to test just my own package?

### CODE REVIEW

**Comment by simonbuchan** [comment]
Note, after figuring out why this seems to break `lambda-tester` types, I might push more changes onto this if it stays open for things like splitting up the tests, and improving them in general, improving event doc links (many have gone out of date and get redirected to generic service doc pages), etc.

**Comment by mrjackyliang** [comment]
@simonbuchan If you have time, mind you add the `Request` key in `APIGatewayEventRequestContext`? Thanks. Here's the issue: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42049

**Comment by simonbuchan** [comment]
Working on splitting (and updating) the tests too... please don't merge for today, at least.

**Comment by simonbuchan** [comment]
Hmm, it's rejecting the use of `include: ["tests"]` in tsconfig.json: what's the motivation there? I'm concerned that a new service gets added with tests, but adding the entry to `tsconfig.json` is missed, since this is a very copy-pastable pattern. I guess just be careful in PR reviews?

**Comment by simonbuchan** [comment]
Mostly squashed the commits, and hopefully match the oddly strict rules on test files now?

**Comment by simonbuchan** [comment]
😕 No idea what's up with those perf changes, there are no significant changes to the actual types, and only somewhat more test code. Can breaking up the test files cause that difference, less caching or something?

**Comment by jessetrinity** [comment]
I don't see a  `///<reference types="node"/>` before the changes. I adding that alone will make perf worse.

**Comment by simonbuchan** [comment]
Thanks, that makes sense. Not strictly needed for that test file, but it's the expected usage, so I think it makes sense to keep.

**Comment by sandersn** [comment]
@simonbuchan I didn't see any other owners sign off yet. Do you want me to wait? I usually like to have at least one other owner sign off on a change to such a popular package.

I can get `npm test` to work for me, and I think the error you see indicates an old version of `types-publisher`. Try `npm install types-publisher` or just refresh your node_modules entirely.

**Comment by simonbuchan** [comment]
@sandersn Oops, yeah, I was still using an old package install. But now I get:

```
Clean data
Clean logs
Clean output
Using local Definitely Typed at C:\code\personal\DefinitelyTyped.
Parsing definitions...
Found 6444 packages.
Parsing in parallel...
Parsing took 22.072 s
Running: git rev-parse --verify master
5185ec9bb068299979201a71c89f08ad6d683e12
Running: git diff master --name-status
Error: RangeError [ERR_CHILD_PROCESS_STDIO_MAXBUFFER] [ERR_CHILD_PROCESS_STDIO_MAXBUFFER]: stdout maxBuffer length exceeded
    at Socket.onChildStdout (child_process.js:363:14)
    at Socket.emit (events.js:321:20)
    at addChunk (_stream_readable.js:297:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:214:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)

    at Object.execAndThrowErrors (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\util\util.ts:160:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at run (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\tester\test-runner.ts:336:24)
    at gitDiff (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\tester\test-runner.ts:324:17)
    at getAffectedPackagesFromDiff (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\tester\test-runner.ts:93:19)
    at Object.runTests [as default] (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\tester\test-runner.ts:81:65)
    at main (C:\code\personal\DefinitelyTyped\node_modules\types-publisher\src\tester\test.ts:34:5)
```

Ehh, not important for this PR at least.

I would really like other sign-off too, especially for such a sweeping change, but despite the large number of listed maintainers, few of them seem to respond to PRs. Not sure what the etiquette is here, but my best guess is if no-one's complained yet, I don't think they will (until this breaks someone)

**Comment by sandersn** [comment]
re being careful in tsconfig: CI checks that all files are either reachable from entries in tsconfig.json or are listed in a file named OTHER_FILES.txt. So if CI passes then you haven't missed any files.

**Comment by sandersn** [comment]
@simonbuchan are you available for a day or so to watch for incoming issues if I merge this? 1.5 million downloads/month is no joke, so I suspect that you will see issues filed if anything breaks.

**Comment by simonbuchan** [comment]
@sandersn Good to know re: tsconfig!

I've been fairly active for this library in terms of PRs at least. The main thing with just issue reports seems to be that the authors seem to need to be manually at-ed. and we have a lot on this library, so I tend not to get on the list. I'll keep an eye on search for at least the next few work days. (I should go through the currently open ones and make PRs where it makes sense...)

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
OK, I merged it. It should publish in the next few minutes.

File: types/aws-lambda/index.d.ts
Comment (apepper) [inline types/aws-lambda/index.d.ts]: I bet that this file should be named `handler` not `hander`.

File: types/aws-lambda/common/api-gateway.d.ts
Comment (apepper) [inline types/aws-lambda/common/api-gateway.d.ts]: That should be a `APIGatewayProxyEvent` not a Praxy event.

**Comment by simonbuchan** [comment]
@apepper, thanks, those damn typos will get you every time. I'll PR these tomorrow.

**Pull request merged**
